crates/ibc/src/core/ics02_client/handler/misbehaviour.rs

```diff
@@ -0,0 +1,80 @@
+//! Protocol logic specific to processing ICS2 messages of type `MsgUpgradeAnyClient`.
+//!
+use crate::core::ics02_client::client_state::ClientState;
+use crate::core::ics02_client::context::ClientReader;
+use crate::core::ics02_client::error::Error;
+use crate::core::ics02_client::events::ClientMisbehaviour;
+use crate::core::ics02_client::handler::ClientResult;
+use crate::core::ics02_client::msgs::misbehaviour::MsgSubmitMisbehaviour;
+use crate::core::ics24_host::identifier::ClientId;
+use crate::events::IbcEvent;
+use crate::handler::{HandlerOutput, HandlerResult};
+use crate::prelude::*;
+
+/// The result following the successful processing of a `MsgSubmitMisbehaviour` message.
+/// This data type should be used with a qualified name `misbehaviour::Result` to avoid ambiguity.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Result {
+    pub client_id: ClientId,
+    pub client_state: Box<dyn ClientState>,
+}
+
+pub fn process(
+    ctx: &dyn ClientReader,
+    msg: MsgSubmitMisbehaviour,
+) -> HandlerResult<ClientResult, Error> {
+    let mut output = HandlerOutput::builder();
+
+    let MsgSubmitMisbehaviour {
+        client_id,
+        misbehaviour,
+        signer: _,
+    } = msg;
+
+    // Read client state from the host chain store.
+    let client_state = ctx.client_state(&client_id)?;
+
+    if client_state.is_frozen() {
+        return Err(Error::client_frozen(client_id));
+    }
+
+    // Read consensus state from the host chain store.
+    let latest_consensus_state = ctx
+        .consensus_state(&client_id, client_state.latest_height())
+        .map_err(|_| {
+            Error::consensus_state_not_found(client_id.clone(), client_state.latest_height())
+        })?;
+
+    let now = ClientReader::host_timestamp(ctx)?;
+    let duration = now
+        .duration_since(&latest_consensus_state.timestamp())
+        .ok_or_else(|| {
+            Error::invalid_consensus_state_timestamp(latest_consensus_state.timestamp(), now)
+        })?;
+
```

Comment:
Not sure we need this. We do the similar checks a bit later for the consensus state at trusted height at each header (which may be different than latest).

Reply:
I would remove it since the trusted state validation is a short cpu distance away in this flow: `check_misbehaviour_and_update_state() -> check_header_and_validator_set()`
But we can also keep it and add a comment. I think is fine either way.

Reply:
Removed it -> https://github.com/cosmos/ibc-rs/pull/215/commits/017b34ab4ebd4c1bfb7b01f791115ecebeadf12c